internal/xds/translator/cors.go

```diff
@@ -0,0 +1,183 @@
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/envoyproxy/gateway/internal/ir"
+)
+
+// patchHCMWithCorsFilter builds and appends the Cors Filter to the HTTP
+// Connection Manager if applicable.
+func patchHCMWithCorsFilter(mgr *hcmv3.HttpConnectionManager, irListener *ir.HTTPListener) error {
+	if mgr == nil {
+		return errors.New("hcm is nil")
+	}
+
+	if irListener == nil {
+		return errors.New("ir listener is nil")
+	}
+
+	if !listenerContainsCorsPolicy(irListener) {
+		return nil
+	}
+
+	// Return early if filter already exists.
+	for _, httpFilter := range mgr.HttpFilters {
+		if httpFilter.Name == wellknown.CORS {
+			// This should not happen since this is the only place where the cors
+			// filter is added in a listener.
+			return fmt.Errorf("cors filter already exists in hcm: %+v", mgr)
+		}
+	}
+
+	corsFilter, err := buildHCMCorsFilter()
+	if err != nil {
+		return err
+	}
+
+	// Ensure the cors filter is the first one in the filter chain.
+	mgr.HttpFilters = append([]*hcmv3.HttpFilter{corsFilter}, mgr.HttpFilters...)
+
+	return nil
+}
+
+// buildHCMCorsFilter returns a Cors filter from the provided IR listener.
+func buildHCMCorsFilter() (*hcmv3.HttpFilter, error) {
+	corsProto := &corsv3.Cors{}
+
+	corsAny, err := anypb.New(corsProto)
+	if err != nil {
+		return nil, err
+	}
+
+	return &hcmv3.HttpFilter{
+		Name: wellknown.CORS,
+		ConfigType: &hcmv3.HttpFilter_TypedConfig{
+			TypedConfig: corsAny,
+		},
+	}, nil
+}
+
+// listenerContainsCorsPolicy returns true if the provided listener has Cors
+// policies attached to its routes.
+func listenerContainsCorsPolicy(irListener *ir.HTTPListener) bool {
+	if irListener == nil {
+		return false
+	}
+
+	for _, route := range irListener.Routes {
+		if route.CorsPolicy != nil {
+			return true
+		}
+	}
+
+	return false
+}
+
+// patchRouteWithCorsConfig patches the provided route with the Cors config if
+// applicable.
+func patchRouteWithCorsConfig(route *routev3.Route, irRoute *ir.HTTPRoute) error {
+	if route == nil {
+		return errors.New("xds route is nil")
+	}
+	if irRoute == nil {
+		return errors.New("ir route is nil")
+	}
+	if irRoute.CorsPolicy == nil {
+		return nil
+	}
+
+	filterCfg := route.GetTypedPerFilterConfig()
+	if _, ok := filterCfg[wellknown.CORS]; ok {
+		// This should not happen since this is the only place where the cors
+		// filter is added in a route.
+		return fmt.Errorf("route already contains cors config: %+v", route)
+	}
+
+	var (
+		allowOrigins              []*matcherv3.StringMatcher
+		allowMethods              string
+		allowHeaders              string
+		exposeHeaders             string
+		maxAge                    string
+		allowCredentials          *wrappers.BoolValue
+		allowPrivateNetworkAccess *wrappers.BoolValue
+	)
+
+	//nolint:gocritic
+	for _, origin := range irRoute.CorsPolicy.AllowOrigins {
+		if origin.Exact != nil {
+			allowOrigins = append(allowOrigins, &matcherv3.StringMatcher{
```

Comment:
can we reuse the existing stringMatcher function here ?https://github.com/envoyproxy/gateway/blob/f203fe51dbcc2b6da3d3d5ea8a1715aad0eb1a1e/internal/xds/translator/route.go#L152